### ISSUE TITLE
Pyright complains about missing argument to pydantic models

### ISSUE THREAD

**Issue by apollo13** [microsoft/pyright#4556]
**Describe the bug**
Pyright assumes that pydantic models can only be initialized via `__init__` which is not true.

**To Reproduce**
Run the code below, the code utilizes pydantic's settings management (https://docs.pydantic.dev/usage/settings/) which allows for initialization via environment variables.

```
No configuration file found.
pyproject.toml file found at /home/florian/dev/test.
Loading pyproject.toml file at /home/florian/dev/test/pyproject.toml
Pyproject file "/home/florian/dev/test/pyproject.toml" is missing "[tool.pyright]" section.
stubPath /home/florian/dev/test/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.291
/home/florian/dev/test/test.py
  /home/florian/dev/test/test.py:11:12 - error: Argument missing for parameter "auth_key" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.928sec
```

**Expected behavior**
No errors from pyright.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
import os

from pydantic import BaseSettings


class Settings(BaseSettings):
    auth_key: str


os.environ["AUTH_KEY"] = "asd"
settings = Settings()
print(settings.auth_key)
```

**VS Code extension or command-line**
pyright command line version 1.1.291

**Comment by erictraut** [comment]
Pyright has no specialized knowledge of pydantic's internals or behaviors.

Pyright is able to support many of pydantic's features through the use of [PEP 681 (dataclass_transform)](https://peps.python.org/pep-0681/). But standard dataclasses don't support initialization through environment variables.

If you want your code to type check with pyright, you could add a `Field` descriptor. Something like this...

```python
class Settings(BaseSettings):
    auth_key: str = Field(default_factory = lambda : os.environ["AUTH_KEY"])
```

**Comment by apollo13** [comment]
Oh, so I guess this means pydantic's use of dataclass_transform is basically wrong in this case?

**Comment by erictraut** [comment]
No, we designed `dataclass_transform` with pydantic in mind. Prior to the introduction of `dataclass_transform`, most of pydantic's functionality didn't work with standards-based type checkers like pyright. Now, the vast majority of pydantic's functionality works. There are still edge cases where pydantic deviates from the behavior of the stdlib dataclass. If you want your code to work with pyright, you'll need to either avoid those features or make some other accommodation in your code.
